File: 03-Prefix_Sum/724-find_pivot_index.py
# ----------------------------------------------------
#  724. Find Pivot Index (19/08/25); JDG - Py Solution
#  Time complexity : O(n)
#  Space complexity: O(1)
#  ----------------------------------------------------

from typing import List

class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        totalSum = sum(nums)

        leftSum = 0
        rightSum = totalSum
        for i in range(len(nums)):
            rightSum -= nums[i]

            if(rightSum == leftSum):
                return i
            leftSum += nums[i]

        return -1

if __name__ == "__main__":
    nums = [1, 7, 3, 6, 5, 6]
    solution = Solution()

    print(solution.pivotIndex(nums))

File: 03-Prefix_Sum/724-find_pivot_index.cpp
/* ----------------------------------------------------
   724. Find Pivot Index (19/08/25); JDG - CPP Solution
   Time complexity : O(n)
   Space complexity: O(1)
   ---------------------------------------------------- */

#include <iostream>
#include <vector>
#include <numeric>

class Solution {
public:
    int pivotIndex(std::vector<int>& nums) {
        int totalSum = std::accumulate(nums.begin(), nums.end(), 0);

        int leftSum = 0;
        int rightSum = totalSum;
        for(int i = 0; i < nums.size(); i++) {
            rightSum -= nums[i];

            if(rightSum == leftSum) {
                return i;
            }
            leftSum += nums[i];
        }
        return -1;
    }
};

int main() {
    std::vector<int> nums = {1, 7, 3, 6, 5, 6};
    Solution solution;

    std::cout << solution.pivotIndex(nums) << '\n';

    return 0;
}